Responder and Child process tevent chain id improvements

justin-stephenson opened the pull request:
This PR adds the following tevent chain ID functionality:

* Add tevent chain ID logic into responders (log messages with tag [CID #])
* Add ability to parse child log files by passing the backend request ID  into child processes and setting the chain ID inside the child.

sss_chain_id getter/setter functions moved into their own source file (remove dependency to tevent), and some small changes to analyzer are needed.

I skipped LDAP, nsupdate, ad renewal, ipa-getkeytab child processes, as they are not tied to a user request and would not benefit from request tracking as far as I know:

alexey-tikhonov commented on src/util/server.c:
```diff
@@ -484,6 +499,8 @@ int server_setup(const char *name, int flags,
         return ENOMEM;
     }
 
+    debug_from_responder = prg_name_is_responder(name);
```
I would delegate this to service implementations.
IMO, name-based logic in generic framework feels error prone / easy to break.

alexey-tikhonov replied:
Or maybe we can have a default "false" set in server_setup() and then "true" in https://github.com/SSSD/sssd/blob/60353300d547f114ca1261e8c130ea72459b572a/src/responder/common/responder_common.c#L1265 (but we need to make sure every responder indeed use this)

justin-stephenson replied:
Made this change as suggested, `sss_process_init()` looks like it is called by each responder

alexey-tikhonov commented on src/util/debug.h:
```diff
@@ -50,6 +50,7 @@ enum sss_logger_t {
 
 extern const char *sss_logger_str[]; /* mapping: sss_logger_t -> string */
 extern const char *debug_prg_name;
+extern bool debug_from_responder;
```
This is only to be able to print "CID" instead of "RID"?
Now I'm not so sure in a choice...

justin-stephenson replied:
That's correct, I did not see any other way in debug.c to determine if the message was sent from a responder, or backend/child

pbrezina replied:
Strictly speaking, you can infer it from `debug_prg_name`. It might be better so set the from in sss_chain_id_setup()?

```c
uint64_t debug_chain_id;
const char * debug_chain_id_fmt;
```

```c
void sss_chain_id_setup(struct tevent_context *ev, const char *fmt);
```

justin-stephenson replied:
Moving the sss_chain_id_setup() call from `sss_process_init` to `server_setup` (from the other review item) means that `debug_from_responder` value has not been overridden/updated by responders yet when sss_chain_id_setup() is reached. Would a proper solution be adding a new function like sss_chain_id_set_fmt that can be called later?

originally this PR was setting debug_from_responder based on `debug_prg_name` but it was suggested that would likely be too fragile.

justin-stephenson replied:
Just to clarify things, there are conflicting review items here, Alexey suggested to move sss_chain_id_setup() to server_setup(), but then we cannot set the chain ID correct format when calling sss_chain_id_setup() because we don't know we are a responder until sss_process_init() is run (which happens after server_setup)

alexey-tikhonov replied:
Another thing worth mentioning, we need to setup `fmt` from child processes as well, where we don't call `sss_chain_id_setup()`, I think.

pbrezina replied:
So perhaps we can keep sss_chain_id_setup() and add sss_chain_id_set_format()? The format is known in server_setup() you just need to pass it as a parameter.

pbrezina replied:
Why? `server_setup` is called from each process' `main`. You can pass the format as additional argument, similar how we already pass the process name.

justin-stephenson replied:
Okay I understand now, thanks for clarification.

justin-stephenson replied:
This is fixed now, child processes are calling `sss_chain_id_set_format` to set the format

alexey-tikhonov commented:
RHBZs:
 - https://bugzilla.redhat.com/show_bug.cgi?id=2013259
 - https://bugzilla.redhat.com/show_bug.cgi?id=2013260
 - https://bugzilla.redhat.com/show_bug.cgi?id=2013262
 - https://bugzilla.redhat.com/show_bug.cgi?id=2013263

justin-stephenson commented:
Now that the RID # is passed to the child log files, `sssctl analyze request show #` will check these child logs and print those RID # matches also. Should this be default behavior, or only when an argument like `--child` is provided to the request show subcommand?

sumit-bose commented:
> Now that the RID # is passed to the child log files, `sssctl analyze request show #` will check these child logs and print those RID # matches also. Should this be default behavior, or only when an argument like `--child` is provided to the request show subcommand?

Hi,

I tend to the latter. The idea of the  tool is to help to find issues. Adding the child logs by default might make this harder especially if the issue is not related to a child process. It might be even worth to be able to switch on only a single child, e.g. `--child=krb5_child`.

Is there a reason you mostly send the id together with other data via stdin and not with a command line option? Personally I would prefer a comand line option because it makes the child process more easy to use manually. E.g. currently you can run `p11_child` manually even in auth mode by calling

    echo -n PIN | /usr/libexec/sssd/p11_child --auth .....

Now additional 8 bytes are expected, which still would allow to call `p11_child` manually but it would be less obvious. The PIN has to be send via stdin, otherwise it would be visible on the command line and so visible for everyone calling `ps`.

bye,
Sumit

alexey-tikhonov commented on src/util/dbg_chain_id.h:
```diff
@@ -0,0 +1,34 @@
+/*
```
Please, add a justification for this split in the commit message.

pbrezina replied:
Maybe do it the other way around and split to `sss_chain_id_tevent.c/h`? Description in commit message is certainly welcome. This would also make it more distinct in the file name itself.

justin-stephenson replied:
I re-did this commit following Pavel's suggestion instead, and added justification to the commit message.

alexey-tikhonov commented on src/providers/krb5/krb5_child.c:
```diff
@@ -3502,6 +3509,9 @@ int main(int argc, const char *argv[])
     }
 
+    /* Restore the chain id */
+    sss_chain_id_set(kr->old_chain_id);
```
What is the reason to bother with `old_chain_id`? (It is zero anyway.)

Wouldn't it be better to include both "krb5_child started.\n" and "krb5_child completed successfully\n"/"krb5_child failed!\n"?

alexey-tikhonov replied:
The same question applies to others `*_child`.

justin-stephenson replied:
The new version fixes these items, old_chain_id is not needed as you mention.

alexey-tikhonov commented on src/providers/proxy/proxy_child.c:
```diff
@@ -493,6 +495,8 @@ int main(int argc, const char *argv[])
         {"id", 0, POPT_ARG_LONG, &id, 0,
          _("Child identifier (mandatory)"), NULL },
+        {"chain-id", 0, POPT_ARG_LONG, &chain_id, 0,
+         _("Tevent chain ID"), NULL },
```
I would add "... for logging purposes".

justin-stephenson replied:
Fixed.

alexey-tikhonov commented on src/p11_child/p11_child_common.c:
```diff
@@ -376,10 +385,16 @@ int main(int argc, const char *argv[])
     return EXIT_SUCCESS;
 fail:
     DEBUG(SSSDBG_CRIT_FAILURE, "p11_child failed!\n");
+    if (old_chain_id != -1) {
```
Implicit signed/unsigned comparison.

alexey-tikhonov commented on src/p11_child/p11_child_common.c:
```diff
@@ -163,6 +171,7 @@ int main(int argc, const char *argv[])
     char *cert_b64 = NULL;
     bool wait_for_card = false;
     char *uri = NULL;
+	uint64_t old_chain_id = -1;
```
While I understand the purpose, still setting unsigned var to `-1` looks a bit weird.

justin-stephenson replied:
This code is now gone in the latest update, so please disregard.

alexey-tikhonov commented on src/p11_child/p11_child_common.c:
```diff
@@ -101,11 +102,14 @@ static int do_work(TALLOC_CTX *mem_ctx, enum op_mode mode, const char *ca_db,
 {
     uint8_t buf[IN_BUF_SIZE];
     ssize_t len;
+    size_t p = 0;
```
What is the purpose of `p`?

justin-stephenson replied:
This code is now gone in the latest update, so please disregard.

alexey-tikhonov commented:
Isn't it required to set `debug_from_responder` in `p11_child`?

justin-stephenson commented:
> 
> Hi,
> 
> I tend to the latter. The idea of the tool is to help to find issues. Adding the child logs by default might make this harder especially if the issue is not related to a child process. It might be even worth to be able to switch on only a single child, e.g. `--child=krb5_child`.

Okay agreed, I went ahead and added a `--child` boolean option defaulting to false to the `request show` subcommand
> 
> Is there a reason you mostly send the id together with other data via stdin and not with a command line option? Personally I would prefer a comand line option because it makes the child process more easy to use manually. E.g. currently you can run `p11_child` manually even in auth mode by calling
> 
> ```
> echo -n PIN | /usr/libexec/sssd/p11_child --auth .....
> ```
> 
> Now additional 8 bytes are expected, which still would allow to call `p11_child` manually but it would be less obvious. The PIN has to be send via stdin, otherwise it would be visible on the command line and so visible for everyone calling `ps`.

Understood, I went ahead and switched to sending the chain ID as a command line argument for each child process. Could you please test the p11 child works as expected now?

> 
> bye, Sumit

justin-stephenson commented:
> Isn't it required to set `debug_from_responder` in `p11_child`?

Yes, fixed.

justin-stephenson commented:
Now that the chain ID is logged in responder log files, should the analyzer tool always grep responder log files for `CID#` in `request show` output? This is simple but it means removing the `--cachereq`  option as those logs would always be included in the output then.

I added a commit for this `Analyzer: Search all responder log files` in case this is agreed upon behavior (will drop it if not). It does produce more output but it would  be useful in solving any responder-logged problems

alexey-tikhonov commented:
Isn't it required to execute `sss_chain_id_setup()` in all child processes?

alexey-tikhonov commented on src/responder/common/responder_common.c:
```diff
@@ -1311,6 +1319,10 @@ int sss_process_init(TALLOC_CTX *mem_ctx,
         goto fail;
     }
 
+    sss_chain_id_setup(rctx->ev);
```
Maybe it makes sense to move this call to `server_setup()` (and remove from `data_provider_be.c::main()`)? Just to have it in a single place.

justin-stephenson replied:
I tried this but it caused some issues with RID# tag missing on all applicable messages in the backend. I'll have to look into this further later if it is necessary.

alexey-tikhonov replied:
While not strictly necessary, would be good to at least understand this, imo.

alexey-tikhonov replied:
Any ideas?

justin-stephenson replied:
I made this change locally and it is working now, but I will wait for responses on the other review items before pushing, as it may be depend on the final solution.

justin-stephenson replied:
This is fixed and pushed

alexey-tikhonov commented on src/responder/common/responder_common.c:
```diff
@@ -549,6 +555,9 @@ static void accept_fd_handler(struct tevent_context *ev,
     }
 
+    /* Set the chain id */
+    cctx->old_chain_id = sss_chain_id_set(rctx->client_id_num);
```
I'm not sure I understand how it works.

If you set chain-id in `accept_fd_handler()` then new requests from all currently connected clients will have the same CID (of last connected client), no?

Wouldn't it be better to set chain-id somewhere "around" `client_fd_handler()`?

justin-stephenson replied:
When setting the chain ID (and incrementing the id number) inside client_fd_handler() functions it causes every new cache request to increment the ID, rendering the client tracking ineffective. That is why incrementing the id number is done in accept_fd_handler() 

Yes there will be a downside of duplciate IDs for any client or daemon which runs multiple commands over a single socket connection, but IIUC we agreed to accept this deficiency for now - knowing that ID tracking is often useful to assist in troubleshooting user run command tests.

alexey-tikhonov replied:
I don't question incrementing in `accept_fd_handler()`.
I question `sss_chain_id_set()` there.

Consider a case of multiple client connected (and making requests) simultaneously - those needs to be distinguished.

justin-stephenson replied:
Okay, I moved `sss_chain_id_set` into `sss_client_fd_handler()`. Thank you

alexey-tikhonov commented on src/util/sss_chain_id.h:
```diff
@@ -21,10 +21,9 @@
-#include <tevent.h>
+#include <stdint.h>
 
-/* Setup chain id tracking on tevent context. */
```
It seems, comment was lost.

justin-stephenson replied:
It should be in `sss_chain_id_tevent.h`

alexey-tikhonov commented on src/providers/ipa/selinux_child.c:
```diff
@@ -258,6 +262,10 @@ int main(int argc, const char *argv[])
         }
     }
 
+    if (chain_id > 0) {
```
What is the reason for this check?

justin-stephenson replied:
Removed.

alexey-tikhonov commented on src/responder/pam/pamsrv_p11.c:
```diff
@@ -746,6 +748,14 @@ struct tevent_req *pam_check_cert_send(TALLOC_CTX *mem_ctx,
+
+    chain_id = sss_chain_id_get();
+
+    if (chain_id > 0) {
```
In `ipa_selinux.c::selinux_fork_child()` this is done unconditionally.
Both variants are fine, but approach feels inconsistent.

justin-stephenson replied:
No need for this conditional in the child processes, it is now removed.

alexey-tikhonov replied:
I meant parent processes here.

justin-stephenson replied:
I thought it was causing unit test failures, but I checked again and it is not an issue. Fixed now.

alexey-tikhonov commented on src/responder/pam/pamsrv_p11.c:
```diff
@@ -718,10 +719,11 @@ struct tevent_req *pam_check_cert_send(TALLOC_CTX *mem_ctx,
     int pipefd_to_child[2] = PIPE_INIT;
     int pipefd_from_child[2] = PIPE_INIT;
-    const char *extra_args[16] = { NULL };
+    const char *extra_args[20] = { NULL };
```
Why not +2?

justin-stephenson replied:
I thought this was an over-estimation of size, but maybe not. I set it to '18'

sumit-bose commented on src/providers/krb5/krb5_child_handler.c:
```diff
@@ -418,6 +420,19 @@ errno_t set_extra_args(TALLOC_CTX *mem_ctx, struct krb5_ctx *krb5_ctx,
+            goto done;
+        }
+    }
+    c++;
```
Hi,

you should call `c++` inside the if-block. However, in most other places you do not check of the returned chain_id is 0 and add the command line option unconditionally. I think it is ok to do the same here.

bye,
Sumit

justin-stephenson replied:
Fixed, thank you.

sumit-bose commented:
Hi,

I'm fine with the patches and test are working locally for me. I restarted the CI task to hopefully get some CI results as well.

bye,
Sumit

sumit-bose commented:
Hi,

I'm sorry I didn't pull the latest version when running the tests. Since the chain_id option is now added unconditionally `test_set_extra_args` needs a change like:

```
diff --git a/src/tests/cmocka/test_krb5_common.c b/src/tests/cmocka/test_krb5_common.c
index c43d836ad..4bf3237a7 100644
--- a/src/tests/cmocka/test_krb5_common.c
+++ b/src/tests/cmocka/test_krb5_common.c
@@ -103,7 +103,8 @@ void test_set_extra_args(void **state)
     assert_int_equal(ret, EOK);
     assert_string_equal(krb5_child_extra_args[0], uid_opt);
     assert_string_equal(krb5_child_extra_args[1], gid_opt);
-    assert_null(krb5_child_extra_args[2]);
+    assert_string_equal(krb5_child_extra_args[2], "--chain-id=0");
+    assert_null(krb5_child_extra_args[3]);
     talloc_free(krb5_child_extra_args);
 
     krb5_ctx->canonicalize = true;
@@ -113,7 +114,8 @@ void test_set_extra_args(void **state)
     assert_string_equal(krb5_child_extra_args[0], uid_opt);
     assert_string_equal(krb5_child_extra_args[1], gid_opt);
     assert_string_equal(krb5_child_extra_args[2], "--canonicalize");
-    assert_null(krb5_child_extra_args[3]);
+    assert_string_equal(krb5_child_extra_args[3], "--chain-id=0");
+    assert_null(krb5_child_extra_args[4]);
     talloc_free(krb5_child_extra_args);
 
     krb5_ctx->realm = discard_const(TEST_REALM);
@@ -124,7 +126,8 @@ void test_set_extra_args(void **state)
     assert_string_equal(krb5_child_extra_args[1], gid_opt);
     assert_string_equal(krb5_child_extra_args[2], "--realm=" TEST_REALM);
     assert_string_equal(krb5_child_extra_args[3], "--canonicalize");
-    assert_null(krb5_child_extra_args[4]);
+    assert_string_equal(krb5_child_extra_args[4], "--chain-id=0");
+    assert_null(krb5_child_extra_args[5]);
     talloc_free(krb5_child_extra_args);
 
     /* --fast-principal will be only set if FAST is used */
@@ -136,7 +139,8 @@ void test_set_extra_args(void **state)
     assert_string_equal(krb5_child_extra_args[1], gid_opt);
     assert_string_equal(krb5_child_extra_args[2], "--realm=" TEST_REALM);
     assert_string_equal(krb5_child_extra_args[3], "--canonicalize");
-    assert_null(krb5_child_extra_args[4]);
+    assert_string_equal(krb5_child_extra_args[4], "--chain-id=0");
+    assert_null(krb5_child_extra_args[5]);
     talloc_free(krb5_child_extra_args);
 
     krb5_ctx->use_fast_str = discard_const(TEST_FAST_STR);
@@ -150,7 +154,8 @@ void test_set_extra_args(void **state)
     assert_string_equal(krb5_child_extra_args[4],
                         "--fast-principal=" TEST_FAST_PRINC);
     assert_string_equal(krb5_child_extra_args[5], "--canonicalize");
-    assert_null(krb5_child_extra_args[6]);
+    assert_string_equal(krb5_child_extra_args[6], "--chain-id=0");
+    assert_null(krb5_child_extra_args[7]);
     talloc_free(krb5_child_extra_args);
 
     krb5_ctx->lifetime_str = discard_const(TEST_LIFE_STR);
@@ -168,7 +173,8 @@ void test_set_extra_args(void **state)
     assert_string_equal(krb5_child_extra_args[6],
                         "--fast-principal=" TEST_FAST_PRINC);
     assert_string_equal(krb5_child_extra_args[7], "--canonicalize");
-    assert_null(krb5_child_extra_args[8]);
+    assert_string_equal(krb5_child_extra_args[8], "--chain-id=0");
+    assert_null(krb5_child_extra_args[9]);
     talloc_free(krb5_child_extra_args);
 
     talloc_free(krb5_ctx);
```

justin-stephenson commented:
> 

Thanks, sorry I should have caught this. Fixed now.

sumit-bose commented:
Hi,

thanks, remaining CI issues are not related, ACK.

bye,
Sumit

alexey-tikhonov commented:
Deferred until we have a conclusion what upstream release this targets.

pbrezina reviewed (commented):
Added some inline comments.

justin-stephenson commented on src/responder/common/cache_req/cache_req_search.c:
```diff
@@ -420,6 +422,11 @@ static errno_t cache_req_search_dp(struct tevent_req *req,
+    /* store the global chain ID and set it back in callback.
+     * Otherwise we lose the chain ID when dp_send_fn is called
+     * (chain ID is set to 0 in sbus_issue_request_done). */
+    state->chain_id = sss_chain_id_get();
```
This comment is incorrect, it was a misunderstanding on my part. The problem I am seeing is that the global chain ID gets reset to 0 leading to some missed 'CID' tagged logs as shown below (`sbus_dispatch` and further). The issue happens somewhere between `cache_req_search_dp()` and `cache_req_search_done()`. It appears that TEVENT_TRACE_AFTER_LOOP_ONCE is called in this codepath which resets the global ID to 0 in `sss_chain_id_trace_loop()`.

 I've spent time trying to figure out the logic of when TEVENT_TRACE_AFTER_LOOP_ONCE is called, but haven't been able to figure out what is the issue here. The chain_id_get and chain_id_set in the cache_req code in this PR works-around this problem, but I would like to understand why this happens.
~~~
(2021-11-30 14:20:23): [nss] [cache_req_search_cache] (0x0400): [CID#1] CR #0: Object [jstephen@redhat.com] was not found in cache
(2021-11-30 14:20:23): [nss] [cache_req_search_dp] (0x0400): [CID#1] CR #0: Looking up [jstephen@redhat.com] in data provider
(2021-11-30 14:20:23): [nss] [sss_dp_get_account_send] (0x0400): [CID#1] Creating request for [redhat.com][0x1][BE_REQ_USER][name=jstephen@redhat.com:-]
(2021-11-30 14:20:23): [nss] [sss_chain_id_trace_loop] (0x1000): [CID#1] TEVENT_TRACE_AFTER_LOOP_ONCE
(2021-11-30 14:20:23): [nss] [sss_chain_id_trace_loop] (0x1000): TEVENT_TRACE_AFTER_LOOP_ONCE
(2021-11-30 14:20:23): [nss] [sbus_dispatch] (0x4000): Dispatching.
(2021-11-30 14:20:23): [nss] [sss_domain_get_state] (0x1000): Domain redhat.com is Active
(2021-11-30 14:20:23): [nss] [cache_req_search_cache] (0x0400): CR #0: Looking up [jstephen@redhat.com] in cache
(2021-11-30 14:20:23): [nss] [cache_req_search_ncache_filter] (0x0400): CR #0: This request type does not support filtering result by negative cache
(2021-11-30 14:20:23): [nss] [cache_req_search_done] (0x0400): CR #0: Returning updated object [jstephen@redhat.com]
(2021-11-30 14:20:23): [nss] [cache_req_create_and_add_result] (0x0400): CR #0: Found 1 entries in domain redhat.com
(2021-11-30 14:20:23): [nss] [sss_chain_id_trace_loop] (0x1000): TEVENT_TRACE_AFTER_LOOP_ONCE
(2021-11-30 14:20:23): [nss] [cache_req_done] (0x0400): CR #0: Finished: Success
~~~

justin-stephenson replied:
@pbrezina Is ths current workaround acceptable, or do you have a better resolution to this? I haven't been able to determine why this occurs, and how to fix it properly.

pbrezina replied:
Remove all changes from cache_req_send.c and apply this diff (ideally as a separate commit):

```diff
diff --git a/src/sbus/request/sbus_request.c b/src/sbus/request/sbus_request.c
index 99e21509f..546a2249e 100644
--- a/src/sbus/request/sbus_request.c
+++ b/src/sbus/request/sbus_request.c
@@ -25,6 +25,7 @@
 
 #include "util/util.h"
 #include "util/dlinklist.h"
+#include "util/sss_chain_id.h"
 #include "sbus/sbus_request.h"
 #include "sbus/sbus_private.h"
 
@@ -575,6 +576,7 @@ struct sbus_outgoing_request_state {
     const char *key;
     struct sbus_connection *conn;
     DBusMessage *reply;
+    uint64_t chain_id;
 };
 
 static errno_t
@@ -620,6 +622,14 @@ sbus_outgoing_request_send(TALLOC_CTX *mem_ctx,
 
     state->conn = conn;
 
+    /**
+     * The message is sent over top level dbus tevent code. This means that the
+     * chain id information is lost and is not restored when we get reply from
+     * dbus. Therefore we need to remember it and restore it manually when this
+     * request is done.
+     */
+    state->chain_id = sss_chain_id_get();
+
     if (key != NULL) {
         state->key = talloc_strdup(state, key);
         if (state->key == NULL) {
@@ -676,6 +686,8 @@ static void sbus_outgoing_request_done(struct tevent_req *subreq)
     req = tevent_req_callback_data(subreq, struct tevent_req);
     state = tevent_req_data(req, struct sbus_outgoing_request_state);
 
+    sss_chain_id_set(state->chain_id);
+
     ret = sbus_message_recv(state, subreq, &state->reply);
     talloc_zfree(subreq);
```

pbrezina commented on src/util/server.c:
```diff
@@ -451,7 +452,8 @@ static const char *get_pid_path(void)
 }
 
-int server_setup(const char *name, int flags,
+int server_setup(const char *name, bool is_responder,
```
It might be better to use 'const char *chain_id_fmt` instead of boolean to make it more generic. Otherwise it looks good to me, thank you.

justin-stephenson replied:
I thought `is_responder` could be re-usable for some other need later on inside `server_setup()`, rather than only for the chain_id_fmt use case.

pbrezina replied:
Ok, I'm not going to insist on this.

pbrezina commented:
https://github.com/SSSD/sssd/pull/5863#discussion_r768040888 + split adding chain id format to its own commit and it is ack from me.

justin-stephenson commented:
Latest push addresses final items requested by Pavel:
*  splits a commit for [adding chain_id](https://github.com/SSSD/sssd/pull/5863/commits/bc5f885dcab2d6e61863bcc5fdc4ee2c3867f063) 
* removed cache req workaround and added commit https://github.com/SSSD/sssd/pull/5863/commits/a671dc01e219c6004ca24faab50a0c0f14580db8 (proper fix)

This introduced an issue where `[ldb]` messages are tagged with CID/RID, which makes the log analyzer output quite unfriendly/ugly.
~~~
(2021-12-13 15:35:32): [be[redhat.com]] [sdap_save_user] (0x0400): [RID#4] Storing info for user jstephen@redhat.com
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] start ldb transaction (nesting: 1)
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] Added timed event "ldb_kv_callback": 0x1895110
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] Added timed event "ldb_kv_timeout": 0x18a79c0
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] Running timer event 0x1895110 "ldb_kv_callback"
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] Destroying timer event 0x18a79c0 "ldb_kv_timeout"
(2021-12-13 15:35:32): [be[redhat.com]] [ldb] (0x10000): [RID#4] Destroying timer event 0x1895110 "ldb_kv_callback"
(2021-12-13 15:35:32): [be[redhat.com]] [sysdb_search_by_name] (0x0400): [RID#4] No such entry
(2021-12-13 15:35:32): [be[redhat.com]] [sysdb_store_user] (0x1000): [RID#4] User jstephen@redhat.com does not exist.
~~~

It could be addressed with the following diff, some other way in the ldb code, or these logs can be simply filtered out in the analyzer code.  What do you suggest @alexey-tikhonov @sumit-bose ?
~~~
diff --git a/src/util/debug.c b/src/util/debug.c
index b54eca7ce..33d6ea049 100644
--- a/src/util/debug.c
+++ b/src/util/debug.c
@@ -354,7 +354,9 @@ void sss_vdebug_fn(const char *file,
                                debug_prg_name, function, level);
 
     if (debug_chain_id > 0) {
-        sss_debug_backtrace_printf(level, debug_chain_id_fmt, debug_chain_id);
+        if (strcmp(function, "ldb") != 0) {
+            sss_debug_backtrace_printf(level, debug_chain_id_fmt, debug_chain_id);
+        }
     }
 
     sss_debug_backtrace_vprintf(level, format, ap);
~~~

alexey-tikhonov commented:
> This introduced an issue where `[ldb]` messages are tagged with CID/RID, which makes the log analyzer output quite unfriendly/ugly.

But this requires log with `SSSDBG_TRACE_LDB` enabled - something that should not really happen much in the wild, right?
Probably we don't have to do anything with it at all?

>      if (debug_chain_id > 0) {
> -        sss_debug_backtrace_printf(level, debug_chain_id_fmt, debug_chain_id);
> +        if (strcmp(function, "ldb") != 0) {

Doesn't look appealing... Imo, even check for `level` would be better, but see above.

justin-stephenson commented:
> > This introduced an issue where `[ldb]` messages are tagged with CID/RID, which makes the log analyzer output quite unfriendly/ugly.
> 
> But this requires log with `SSSDBG_TRACE_LDB` enabled - something that should not really happen much in the wild, right? Probably we don't have to do anything with it at all?

You are right, I forgot i had debug_level = 10 set. Probably it can be ignored then.
> 
> ```
> >      if (debug_chain_id > 0) {
> > -        sss_debug_backtrace_printf(level, debug_chain_id_fmt, debug_chain_id);
> > +        if (strcmp(function, "ldb") != 0) {
> ```
> 
> Doesn't look appealing... Imo, even check for `level` would be better, but see above.

Ok disregard.

justin-stephenson commented:
These changes remove the analyzer `--cachereq'`option and add a new `--child` option,  https://sssd.io/troubleshooting/analyzer.html will need to be updated when these changes are released.

alexey-tikhonov commented:
@sumit-bose, do you agree with the latest changes?

sumit-bose commented:
Hi,

thanks for the patch, it is working well in my (quite basic) tests for PAM and NSS.

I'm afraid I can't remember what we planned for the other responders, e.g. ssh. You added support in the responders already here but `sssctl analyze` currently has no options to select other responders. Nevertheless, this should not be added here but better with another patch on top of your patch with replaces python click. So, ACK.

bye,
Sumit

alexey-tikhonov commented:
@justin-stephenson, please rebase.

justin-stephenson commented:
> @justin-stephenson, please rebase.

Done, thank you.

justin-stephenson commented:
Latest push adds the following into this PR, to incorporate changes from PR 5845 and avoid the need to rebase

 * Commit from @stanislavlevin in  https://github.com/SSSD/sssd/pull/5845
 * https://github.com/SSSD/sssd/pull/5863/commits/bf38817478bf234adab6043b9ce60e45a76908b6 on top to address circular import issue exposed by PR 5845

alexey-tikhonov commented:
@stanislavlevin, are you fine with a cherry-pick and a fix on top?

stanislavlevin commented:
```
[root@master1 /]# sssctl analyze request list -v
NOTE: Tevent chain ID support missing, request analysis will be limited.
It is recommended to use the --logdir option against tevent chain ID supported SSSD logs.
2021-12-17  8:09:10: [uid 0] CID #: getent
   - User by name
Traceback (most recent call last):
  File "/usr/libexec/sssd/sss_analyze", line 5, in <module>
    sss_analyze.run()
  File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 104, in run
    analyzer.main()
  File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 99, in main
    args.func(args)
  File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 231, in list_requests
    self.print_formatted(line, args.verbose)
  File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 168, in print_formatted
    cr_field = fields[3][7:]
IndexError: list index out of range
Command '/usr/libexec/sssd/sss_analyze' failed with [1]
```

stanislavlevin commented:
cherry-pick is OK.

stanislavlevin commented:
Are there corresponding tests? Can't find them.

stanislavlevin commented:
> ```
> [root@master1 /]# sssctl analyze request list -v
> NOTE: Tevent chain ID support missing, request analysis will be limited.
> It is recommended to use the --logdir option against tevent chain ID supported SSSD logs.
> 2021-12-17  8:09:10: [uid 0] CID #: getent
>    - User by name
> Traceback (most recent call last):
>   File "/usr/libexec/sssd/sss_analyze", line 5, in <module>
>     sss_analyze.run()
>   File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 104, in run
>     analyzer.main()
>   File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 99, in main
>     args.func(args)
>   File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 231, in list_requests
>     self.print_formatted(line, args.verbose)
>   File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 168, in print_formatted
>     cr_field = fields[3][7:]
> IndexError: list index out of range
> Command '/usr/libexec/sssd/sss_analyze' failed with [1]
> ```

```
(Pdb) fields
['(2021-12-17  8:09:10): ', 'nss] ', 'cache_req_search_send] (0x0400): CR #135: Looking up _sssd@ipa.test\n']
```

justin-stephenson commented:
> ```
> [root@master1 /]# sssctl analyze request list -v
> NOTE: Tevent chain ID support missing, request analysis will be limited.
> It is recommended to use the --logdir option against tevent chain ID supported SSSD logs.
> 2021-12-17  8:09:10: [uid 0] CID #: getent
>    - User by name
> Traceback (most recent call last):
>   File "/usr/libexec/sssd/sss_analyze", line 5, in <module>
>     sss_analyze.run()
>   File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 104, in run
>     analyzer.main()
>   File "/usr/lib/python3/site-packages/sssd/sss_analyze.py", line 99, in main
>     args.func(args)
>   File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 231, in list_requests
>     self.print_formatted(line, args.verbose)
>   File "/usr/lib/python3/site-packages/sssd/modules/request.py", line 168, in print_formatted
>     cr_field = fields[3][7:]
> IndexError: list index out of range
> Command '/usr/libexec/sssd/sss_analyze' failed with [1]
> ```

@stanislavlevin Does your distribution provide a recent [libtevent](https://tevent.samba.org/) package ?

The analyzer tool is effectively useless without tevent chain ID support. We do not handle this case well inside analyzer code, perhaps we just return gracefully and print a stronger error message in this situation.

stanislavlevin commented:
@justin-stephenson, I saw the message, thank you.
Since there is no strict requirement on that version of libtevent I assumed the tool should work in either case. Otherwise, this case should be explicitly handled I think.

alexey-tikhonov commented:
From my point of view, it's ok to change "will be limited" to "unsupported" and just quit at this point.

There is really very little we can do without req-ids.

sumit-bose commented:
Hi,

I agree with @alexey-tikhonov, it is better to just fail hard if chain ID support is missing.

bye,
Sumit

justin-stephenson commented:
* Added one final commit (I hope) https://github.com/SSSD/sssd/pull/5863/commits/8ab51e9dbd3ec7a469f0e38046f2354297e7b7ba

alexey-tikhonov commented:
https://s3.eu-central-1.amazonaws.com/sssd-ci/PR-5863/20/fedora33/ci-make-intgcheck.log

justin-stephenson commented:
> https://s3.eu-central-1.amazonaws.com/sssd-ci/PR-5863/20/fedora33/ci-make-intgcheck.log

Not sure how this was caused by my PR, I'm triggering a re-run of the CI before I leave for the weekend. I'll have to investigate further on Monday if it shows the same failures. I couldn't reproduce this failure locally but I am on Fedora 34, not 33

alexey-tikhonov commented:
> > https://s3.eu-central-1.amazonaws.com/sssd-ci/PR-5863/20/fedora33/ci-make-intgcheck.log
> 
> Not sure how this was caused by my PR

I also don't understand this.
But other PRs don't have this fail.

alexey-tikhonov commented:
The same test failed again.
Looks a little bit similar to fails we saw in #5766 (that were fixed by 29f8a795f67db651e95842cddf153975f97d9b3f)

alexey-tikhonov commented on src/util/debug.c:
```diff
@@ -292,15 +292,18 @@ void sss_vdebug_fn(const char *file,
-        if (debug_chain_id > 0) {
+        if (debug_chain_id > 0 && debug_chain_id_fmt != NULL) {
+            strncpy(combined_fmt, debug_chain_id_fmt, sizeof(combined_fmt) - 1);
+            strcat(combined_fmt, "%s");
```
This is actually not safe. Result of `strncpy()` might be non null-terminated or "-1" might be not enough to hold additonal "%s". See below.

alexey-tikhonov replied:
More important issue is that this `strncpy+strcat` will happen every DEBUG message with the same args every time - performance hit.

alexey-tikhonov replied:
`combined_fmt` should be static and initialized once.

We could also use just:
```
#define DEBUG_CHAIN_ID_FMT_RID "[RID#%lu] %s"
#define DEBUG_CHAIN_ID_FMT_CID "[CID#%lu] %s"
```
  --  You already make an assumption about specifiers inside `debug_chain_id_fmt` anyway.

But this would require `sss_debug_backtrace_printf(level, debug_chain_id_fmt, debug_chain_id, "");`...

justin-stephenson replied:
Okay thanks for the comments, I switched to using `snprintf` as proposed and `combined_fmt` is now static

alexey-tikhonov commented on src/util/server.c:
```diff
@@ -472,6 +474,12 @@ int server_setup(const char *name, int flags,
     char *pidfile_name;
     int cfg_debug_level = SSSDBG_INVALID;
 
+    if (is_responder) {
```
This won't compile as `is_responder` arg added in a later commit...

justin-stephenson replied:
Fixed.

alexey-tikhonov commented on src/util/server.c:
```diff
@@ -703,6 +711,8 @@ int server_setup(const char *name, int flags,
         }
     }
 
+    sss_chain_id_setup(ctx->event_ctx);
```
IIUC, this and `sss_chain_id_set_format()` should be moved to 6a54a91beb8767bd3e7df5ce61c128cd985e4feb

justin-stephenson replied:
Fixed

alexey-tikhonov commented on src/util/debug.c:
```diff
@@ -53,6 +51,7 @@ enum sss_logger_t sss_logger = STDERR_LOGGER;
 const char *debug_log_file = "sssd";
 FILE *_sss_debug_file;
 uint64_t debug_chain_id;
+const char *debug_chain_id_fmt = "BUG%lu";
```
Could you please add a comment like "Default value says 'BUG' because this is just a precautionary measure and it is expected value is always set explicitly by every SSSD process. 'BUG' should make any potential oversight more prominent in the logs."

justin-stephenson replied:
Done

alexey-tikhonov commented:
Pushed PR: https://github.com/SSSD/sssd/pull/5863

* `master`
    * 1110bd59e62dc9a9886612f59eb6358fa53fb0cc - Analyzer: Fail if chain ID support is missing
    * 9d62708176ce1c9d50f8f83f18497321aa86b6ae - Analyzer: Avoid circular import
    * ca1d7e29103c5720a639ae5b734846077e097d05 - sss-analyze: Fix self imports
    * 7825e0d3223a374d002d0feb23441f0a8f8687f7 - Analyzer: Search all responder log files
    * 0ba456f9edb978bb7b28452a8d3afa32fdba66cc - Analyzer: Add --child argument to 'request show'
    * 1959a2bb47c6f4e08f5bcddb4697e2ddde719c42 - Analyzer: Parse the responder request ID
    * 06d3e79c3baba5f9eae8f5c814a0b47ccc8856a7 - proxy_child: Add chain ID logging support
    * be482ac39075ce7d9112dfaeb962d2a01ba2a910 - p11_child: Add chain ID logging support
    * c92d39a30fa0162d4efdfbe5883c8ea9911a2249 - ipa_selinux: Add chain ID logging support
    * 60712f31fbc17fe2fd49364327c18e5da2697016 - gpo: Add chain ID logging support
    * cb70739f31f4051f52af03e1e2ba1e079689a231 - krb5_child: Add chain ID logging support
    * 526f731495c5a7fd13309a41bad61b4f95c5d2b4 - chain_id: Add support for custom debug format
    * 2b6edf7733fc69b7c7fb79ee935e77a9a1ec181d - RESPONDER: Support chain ID logging
    * 9296eaf9af04503807a0015e0405e2ff651ee532 - sbus: Remember outgoing request chain ID
    * 6f217eac888bb25c875f1bb060f6cef5a052e2f3 - RESPONDER: Remove extraneous client ID logging
    * e8e7e23afeb5f07c82dcb05341f26f92ee548ab7 - util: Split chain ID tevent functions
